Translate "Passing functions to components" (#95)
reactjs/fr.reactjs.org: content/docs/faq-functions.md

}
```

### Comment éviter qu'une fonction soit appelée trop tôt ou trop de fois ? {#how-can-i-prevent-a-function-from-being-called-too-quickly-or-too-many-times-in-a-row}

Si vous avez un gestionnaire d'événement comme `onClick` ou `onScroll` et que vous voulez éviter que la fonction de rappel soit appelée trop vite, vous pouvez limiter le taux auquel la fonction est exécutée. Vous pouvez le faire en utilisant :

- **le throttling** : changements basés sur une fréquence de temps (exemple : [`_.throttle`](https://lodash.com/docs#throttle))

> - **le debouncing** : publier des changements après une période d'inactivité (exemple : [`_.debounce`](https://lodash.com/docs#debounce))

Reviewer: ```suggestion
- **le _debouncing_** : application des modifications après une période d'inactivité (exemple : [`_.debounce`](https://lodash.com/docs#debounce))
```

> - **le throttling basé sur `requestAnimationFrame`** : changements basés sur [`requestAnimationFrame`](https://developer.mozilla.org/fr/docs/Web/API/Window/requestAnimationFrame) (exemple : [`raf-schd`](https://github.com/alexreardon/raf-schd))

Reviewer: ```suggestion
- **le _throttling_ basé sur `requestAnimationFrame`** : limitation de la fréquence de déclenchement grâce à [`requestAnimationFrame`](https://developer.mozilla.org/fr/docs/Web/API/Window/requestAnimationFrame) (exemple : [`raf-schd`](https://github.com/alexreardon/raf-schd))
```

> Voir [cette exemple](http://demo.nimius.net/debounce_throttle/) pour une comparaison des fonctions `throttle` eet `debounce`.

Reviewer: ```suggestion
Jetez un coup d'œil à [cette visualisation](http://demo.nimius.net/debounce_throttle/) pour une comparaison des fonctions `throttle` et `debounce`.
```

> > Note :

Reviewer: ```suggestion
> Remarque
```

Trad normalisée

>

> > `_.debounce`, `_.throttle` eet `raf-schd` fournissent une méthtode `cancel` pour annuler l'appel à la fonction de rappel. Il est recommandé d'appeler cette méthode dans `componentWillUnmount` _ou_ de s'assurer que le composant est toujours monté dans la fonction retardée.

Reviewer: ```suggestion
> `_.debounce`, `_.throttle` et `raf-schd` fournissent une méthode `cancel` pour annuler l'appel à la fonction de rappel différée. Il est recommandé d'appeler cette méthode dans `componentWillUnmount` _ou_ de s'assurer dans la fonction différée que le composant est toujours monté.
```

#### Throttle {#throttle}

> Le « throttling » évite qu'une fonction ne soit appelée plus d'une fois dans un certain laps de temps. L'exemple ci-dessous retarde le gestionnaire d'événément de « click » pour éviter qu'il soit appelé plus d'une fois par seconde.

Reviewer: ```suggestion
Le _throttling_ évite qu'une fonction ne soit appelée plus d'une fois dans un certain laps de temps. L'exemple ci-dessous limite le gestionnaire d'événements de clic pour éviter qu'il ne soit appelé plus d'une fois par seconde.
```

  }

  render() {

>     return <button onClick={this.handleClickThrottled}>Charger plus</button>;

Reviewer: ```suggestion
    return <button onClick={this.handleClickThrottled}>Charger la suite</button>;
```

#### Debounce {#debounce}

> Le « debouncing » assure qu'une fonction ne sera appelée qu'après un certain temps soit passé après le dernier appel à cette fonction. Cela peut être utile quand vous avez un calcul complexe à faire après un événement qui peut être déclenché rapidement (comme le défilement d'une page ou les frappes d'un clavier). L'exemple ci-dessous utilise cette méthode sur un champ de texte avec un délai de 250ms.

Reviewer: ```suggestion
Le _debouncing_ garantit qu'une fonction ne sera appelée qu'après qu‘un certain temps a passé depuis le dernier appel à cette fonction : on garantit un intervalle minimal entre deux exécutions. Ça peut être utile quand vous avez un calcul complexe à faire en réponse à un événement susceptible d’être déclenché fréquemment (comme le défilement d'une page ou les frappes au clavier). L'exemple ci-dessous utilise cette méthode sur un champ texte avec un délai de 250ms.
```

      <input
        type="text"
        onChange={this.handleChange}
        placeholder="Recherche..."
        defaultValue={this.props.value}
      />
    );
  }

  handleChange(e) {

>     // React regroupe les événements, on a donc besoin de lire la valeur avec le debouce

Reviewer: ```suggestion
    // React recycle les événements, on a donc besoin de lire la valeur avant le différé.
```

Glissement sémantique

>     // Sinon, on peut appler `event.persist()` et passer l'intégralité de l'événement

Reviewer: ```suggestion
    // On aurait aussi pu appeler `event.persist()` et passer l'événement complet.
```

>     // Pour plus d'information, allez voir fr.reactjs.org/docs/events.html#event-pooling

Reviewer: ```suggestion
    // Pour en apprendre davantage, consultez fr.reactjs.org/docs/events.html#event-pooling
```

}
```

> #### le « throttling » de `requestAnimationFrame` {#requestanimationframe-throttling}

Reviewer: ```suggestion
#### _Throttling_ basé sur `requestAnimationFrame` {#requestanimationframe-throttling}
```

> [`requestAnimationFrame`](https://developer.mozilla.org/fr/docs/Web/API/Window/requestAnimationFrame) est un moyen de mettre en file d'attente une fonction pour qu'elle soit exécutée par le navigateur à un moment optimal pour les performances d'affichage. Une fonction mise en file d'attente avec `requestAnimationFrame` sera exécutée dans la prochaine image (*frame*, NdT). Le navigateur s'assurera du mieux qu'il peut qu'il y'ait toujours 50 images par secondes (60 ips). Cependant, si le navigateur en est incapable, il limitera naturellement le nombre d'images par seconde. Par exemple, un appareil pourrait n'être capable d'afficher que 30 images par seconde, vous n'obtiendrez donc que 30 images par seconde. Utilisez `requestAnimationFrame` pour faire du « throttling » est une technique pratique afin d'éviter de faire plus de 60 mises à jour par seconde. Si vous faites 100 mises à jour en une seconde, vous créez une charge de travail supplémentaire pour le navigateur que l'utilisateur ne pourra pas voir.

Reviewer: ```suggestion
[`requestAnimationFrame`](https://developer.mozilla.org/fr/docs/Web/API/Window/requestAnimationFrame) permet de différer une fonction pour qu'elle soit exécutée par le navigateur à un moment optimal pour les performances d'affichage. Une fonction différée avec `requestAnimationFrame` sera exécutée à la prochaine passe d'affichage *(frame, NdT)*. Le navigateur fera de son mieux pour qu'il y'ait toujours 60 passes par seconde (60 fps). Cependant, si le navigateur n’y arrive pas, il _limitera_ naturellement le nombre de passes par seconde. Par exemple, un appareil pourrait n'être capable d'afficher que 30 fps, vous n'obtiendrez donc que 30 passes par seconde. Utiliser `requestAnimationFrame` pour limiter l’exécution est une technique pratique afin d'éviter de faire plus de 60 mises à jour par seconde. Si vous faites 100 mises à jour en une seconde, vous créez une charge de travail supplémentaire pour le navigateur que l'utilisateur ne pourra de toutes façons pas voir.
```

> >**Note :**

Reviewer: ```suggestion
>Remarque
```

Trad normalisée (terme, pas de deux-points) et graisse normalisée (la CSS auto-graisse la première ligne des notes depuis un bail, les anciennes graisses manuelles rendent moches en surcouche du coup)

>

> >Utilisez cette technique ne capturera que la dernière valeur publiée pour une image donnée. Vous pouvez voir un exemple de comment marche cette optimisation sur [`MDN`](https://developer.mozilla.org/fr/docs/Web/Events/scroll).

Reviewer: ```suggestion
>Utiliser cette technique ne capturera que la dernière valeur publiée à chaque passe. Vous pouvez voir un exemple du fonctionnement de cette optimisation sur le [`MDN`](https://developer.mozilla.org/fr/docs/Web/Events/scroll).
```

    this.handleScroll = this.handleScroll.bind(this);

    // Crée une nouvelle fonction à planifier
    this.scheduleUpdate = rafSchedule(
      point => this.props.onScroll(point)
    );
  }

  handleScroll(e) {

>     // Quand on reçoit un événement de défilement de la page, planifier une mise à jour

Reviewer: ```suggestion
    // Quand on reçoit un événement de défilement de la page, planifier une mise à jour.
```

>     // Si on reçoit plusieurs mises à jour dans cette image, on ne publiera que la dernière valeur.

Reviewer: ```suggestion
    // Si on reçoit plusieurs mises à jour dans la même passe, on ne publiera que la dernière valeur.
```

    this.scheduleUpdate({ x: e.clientX, y: e.clientY });
  }

  componentWillUnmount() {

>     // Annule toute mise à jour en attente puisqu'on démonte le composant (unmount, NdT)

Reviewer: ```suggestion
    // Annule toute mise à jour en attente puisqu'on démonte le composant
```

La NdT est superflue à ce stade, le reste de la doc pose bien le vocabulaire, et puis la VO n'apporte pas de clarification ici

}
```

> #### Tester votre limitation de débit {#testing-your-rate-limiting}

Reviewer: ```suggestion
#### Tester votre limitation de fréquence {#testing-your-rate-limiting}
```

Dans le cadre des appels de fonctions, on parle de fréquence, pas de débit (plus un terme pour les flux)

> Quand vous testez si votre code de limitation de débit fonctionne correctement, il est utile de pouvoir manipuler le temps. Si vous utilisez [`jest`](https://facebook.github.io/jest/), vous pouvez utiliser les [`faux minuteurs`](https://facebook.github.io/jest/docs/en/timer-mocks.html) pour accélérer le temps. Si vous utiliser `requestAnimationFrame`, vous pouvez utiliser [`raf-stub`](https://github.com/alexreardon/raf-stub) afin de contrôler les images d'animation.

Reviewer: ```suggestion
Afin de tester que votre code de limitation de fréquence fonctionne correctement, il est utile de pouvoir manipuler le temps. Si vous utilisez [`jest`](https://facebook.github.io/jest/), vous pouvez utiliser les [fausses horloges](https://facebook.github.io/jest/docs/en/timer-mocks.html) *(mock timers, NdT)* pour accélérer le temps. Si vous utilisez `requestAnimationFrame`, vous pouvez utiliser [`raf-stub`](https://github.com/alexreardon/raf-stub) afin de contrôler la progression des passes d'animation.
```
